### PR TITLE
[12.0][FIX] Issue 187: Expected Singleton

### PR DESCRIPTION
**Pull request by bodedra**
Issue Reference : #187

### CODE REVIEW

**Comment by bodedra** [comment]
Would you please review these.

@max3903 @brian10048 @kittiu

**Comment by kittiu** [comment]
@bodedra I have test it. There is no error now. 

image

But with the above code, all fsm.order in self would be written with the same customer_id (from last order), when they might not be?

Or should it be something like,

image

**Comment by bodedra** [comment]
@kittiu May you test it again ? I have improve logic.

If these approach will fail, will try with yours.

**Comment by kittiu** [comment]
I think it loop through orders but still write in self multiple time? I am not sure that would work.

**Comment by bodedra** [comment]
@kittiu Odoo hr_recruitment module has issued same technique . Here is reference https://github.com/odoo/odoo/blob/12.0/addons/hr_recruitment/models/hr_recruitment.py#L284

**Comment by kittiu** [comment]
@bodedra yes, I have seen it. Is it possible that the hr_recruitment is also not correct? But when it try to update **last_stage_id** may be it is ok as to use the stage from the last applicant?
Same here, I am not sure it is ok to write using value from the last fsm.order? (I really don't know what is expected in term functional)

Note: I think within the for loop it will be using the value from cache to write to the records. If that so, it will be equal with write to all orders with the value from the last order?

May be we need other to have second opinion?

**Comment by bodedra** [comment]
@kittiu That's my bad. Please ignore hr_recruitment. :smile:  Yes. It updates last location customer in all FSM order. I have tried locally with your suggested code and work fine. :+1:

**Comment by max3903** [comment]
@bodedra 

```
@api.multi
def write(self, vals)
    for order in self:
        ...
    return super(FSMOrder, self).write(vals)
```

**Comment by bodedra** [comment]
> @bodedra
> 
> ```
> @api.multi
> def write(self, vals)
>     for order in self:
>         ...
>     return super(FSMOrder, self).write(vals)
> ```

@max3903 These will not take updated FSM location. So IMO our current fix will be best for updated location with their customer.

File: fieldservice/models/fsm_order.py [hunk 1]
```diff
@@ -199,9 +199,11 @@ def write(self, vals):
                 vals.get('scheduled_date_start')) + \
                 timedelta(hours=self.scheduled_duration)
             vals['scheduled_date_end'] = str(date_to_with_delta)
+        res = super(FSMOrder, self).write(vals)
         if 'customer_id' not in vals:
-            vals['customer_id'] = self.location_id.customer_id.id
-        return super(FSMOrder, self).write(vals)
+            for fsm_order in self:
+                fsm_order.customer_id = fsm_order.location_id.customer_id.id
+        return res
```
Comment (max3903) [inline fieldservice/models/fsm_order.py:206]:
```
        res = super(FSMOrder, self).write(vals)
        for order in self:
            if 'customer_id' not in vals and order.customer_id is False:
                order.customer_id = order.location_id.customer_id.id
        return res
```

**Comment by bodedra** [comment]
@max3903 Thanks for your suggestion. Comment attended!

**Review by max3903 — APPROVED** [review]

**Review by brian10048 — APPROVED** [review]

**Pull request merged**